okhttp/src/main/java/com/squareup/okhttp/OkHttpTlsUpgrader.java

```diff
@@ -69,21 +74,27 @@ public static SSLSocket upgrade(SSLSocketFactory sslSocketFactory,
     spec.apply(sslSocket, getOkHttpRoute(host, port, spec));
 
     Platform platform = Platform.get();
-    try {
-      // Force handshake.
-      sslSocket.startHandshake();
-
-      String negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
-      if (negotiatedProtocol == null) {
-        throw new RuntimeException("protocol negotiation failed");
+    // It's possible that the user provided SSLSocketFactory has already done the handshake
```

Comment:
Okay, that's good to know. That makes me all the more certain that sslSocket.getSession() is what we want. sslSocket.startHandshake() will start a new handshake if the first has already completed, which may have been done without ALPN/NPN support. Not that that case matters much, since we will end up failing anyway, but the code is more straightforward if we always call sslSocket.getSession() to wait until the handshake is done.